lib/prelude.ml

```diff
@@ -110,3 +110,106 @@ instance monad list begin
   let xs >>= f = [ y | with x <- xs, with y <- f x ]
   let join xss = [ x | with xs <- xss, with x <- xs ]
 end
+
+type option 'a = Some of 'a | None
+
+let from_option x = function
+  | Some x -> x
+  | None -> x
+
+instance eq 'a => eq (option 'a) begin
+  let x == y =
+    match x, y with
+    | None, None -> true
+    | Some x, Some y -> x == y
+    | _, _ -> false
+end
+
+instance functor option begin
+  let f <$> x =
+    match x with
+    | None -> None
+    | Some x -> Some (f x)
+end
+
+instance applicative option begin
+  let pure = Some
+  let f <*> x =
+    match f, x with
+    | Some f, Some x -> Some (f x)
+    | _, _ -> None
+end
+
+instance monad option begin
+  let xs >>= f =
+    match xs with
+    | Some x -> f x
+    | None -> None
+end
+
+external val parse_int : string -> option int =
+  "function(x) \
+     if tonumber(x) and ((math.modf(tonumber(x))) == tonumber(x)) then \
+       return Some(tonumber(x)) \
```

Comment:
So strictly speaking, this may not be safe - `Some`/`None` may be renamed. I'm not 100% sure it'll even be emitted even - only if people actually construct values using it (though would need to check).

I think the "cleaner" thing to do here would be to have some `Obj.magic`-esque function and expose `type` - then we can have some function from `lua_repr -> option 'a` (clearly unsound, but hopefully used in a safe manner).

Reply:
I can make it `private` then pass the constructors to it, which is a better stopgap solution here if you ask me

Reply:
Ohh, that's simple and clean. 👍